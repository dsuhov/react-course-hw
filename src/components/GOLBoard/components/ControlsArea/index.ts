export * from "./ControlsArea";
export * from "./saga";
