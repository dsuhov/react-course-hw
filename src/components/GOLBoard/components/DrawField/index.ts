export * from "./DrawField";
export * from "./components";
