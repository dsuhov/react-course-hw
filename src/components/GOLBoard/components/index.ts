export * from "./DrawField";
export * from "./ControlsArea";
export * from "./StatusLine";
