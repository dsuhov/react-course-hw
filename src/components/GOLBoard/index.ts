export * from "./GOLBoard";
export * from "./components";
