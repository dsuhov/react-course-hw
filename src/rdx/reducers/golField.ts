import { Action } from "redux";
import { clearField } from "@/utils/clearField";
import { getUpdatedField } from "@/utils/getUpdatedField";
import { fillField } from "@/utils/fillField";
import * as actionTypes from "@/rdx/actions/actionTypes";

const defaultFieldScheme: FieldScheme = clearField([10, 10]);

const getFieldCellClick = (
  state: FieldScheme,
  coords: CellCorrds
): FieldScheme => {
  const { x, y } = coords;
  const isXValid = x >= 0 && x < state[0].length;
  const isYValid = y >= 0 && y < state.length;
  const areCoordinatesValid = isXValid && isYValid;
  if (!areCoordinatesValid) {
    return state;
  }

  const newField = state.map((row) => [...row]);
  newField[y][x] = !state[y][x];

  return newField;
};

export const golField = (
  state: FieldScheme = defaultFieldScheme,
  action: Action & { payload?: any }
): FieldScheme => {
  switch (action.type) {
    case actionTypes.UPDATE_FIELD:
      return getUpdatedField(state);
    case actionTypes.CLEAR_FIELD:
      return clearField([state[0].length, state.length]);
    case actionTypes.FILL_FIELD:
      return fillField(
        action.payload.fullness,
        action.payload.size[0],
        action.payload.size[1]
      );
    case actionTypes.CELL_CLICK:
      return getFieldCellClick(state, action.payload);
  }

  return state;
};
