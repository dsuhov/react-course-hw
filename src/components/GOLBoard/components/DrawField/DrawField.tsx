import React, { FC } from "react";
import styled from "@emotion/styled";
import { Cell } from "./components";

interface DrawFieldProps {
  fieldScheme: boolean[][];
  cellClickHandler: (x: number, y: number) => void;
}

interface FieldWrapperProps {
  size: number;
}

const FieldWrapper = styled.div<FieldWrapperProps>`
  display: flex;
  flex-wrap: wrap;
  ${({ size }: FieldWrapperProps) => `width: ${size * 11 + 1}px`};
  margin: 0 auto;
  border: 5px solid #cdc392;
  box-shadow: 0 0 10px #304c89;
  border-radius: 2px;
`;

export const DrawField: FC<DrawFieldProps> = ({
  fieldScheme,
  cellClickHandler,
}) => {
  return (
    <FieldWrapper size={fieldScheme[0].length}>
      {fieldScheme.map((row, y) => [
        ...row.map((status: boolean, x) => (
          <Cell
            key={`${x}_${y}`}
            x={x}
            y={y}
            clickHandler={cellClickHandler}
            isAlive={status}
          />
        )),
      ])}
    </FieldWrapper>
  );
};
