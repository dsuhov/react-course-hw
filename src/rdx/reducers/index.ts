import { golField } from "./golField";
import { gameStatus } from "./gameStatus";
import { combineReducers } from "redux";

export const reducer = combineReducers({
  golField,
  gameStatus,
});
