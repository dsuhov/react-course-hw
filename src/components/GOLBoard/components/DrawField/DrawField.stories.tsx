import React from "react";
import { DrawField } from "./DrawField";
export default {
  title: "Draw Field",
};

import { Provider } from "react-redux";
import { store } from "@/rdx/store";

export const fiel = () => (
  <Provider store={store}>
    <DrawField />
  </Provider>
);
