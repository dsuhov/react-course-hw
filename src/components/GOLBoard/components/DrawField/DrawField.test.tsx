import React from "react";
import { mount, ReactWrapper } from "enzyme";
import renderer from "react-test-renderer";
import { DrawField } from "./DrawField";
import { Cell } from "./components";
import { Provider } from "react-redux";
import { golField as reducer } from "@/rdx/reducers/golField";
import { createStore, Reducer } from "redux";

function newTestScheme(
  xSize: number,
  ySize: number,
  value: boolean
): boolean[][] {
  return (
    Array.from({ length: ySize }).map(() =>
      Array.from({ length: xSize }).fill(value)
    ) as boolean[][]
  );
}

/**
 * Main sense of this test is to check that DrawField component
 * correctly renders correct amount of cells in correct configureation and
 * passes correct data to them.
 */

const setupWrapper = (
  reducer: Reducer<FieldScheme>,
  initState: FieldScheme
): ReactWrapper => {
  const rootState =
    {
      golField: initState,
      gameStatus: { status: "stopped" },
    } as any;

  const newStore = createStore(reducer, rootState);
  const field = mount(
    <Provider store={newStore}>
      <DrawField />
    </Provider>
  );

  return field;
};

describe("Field that draw cells", () => {
  const emptyFieldScheme = newTestScheme(4, 5, false) as FieldScheme;
  const fullFieldScheme = newTestScheme(4, 5, true) as FieldScheme;

  it("reders full empty field", () => {
    const wrapper = setupWrapper(reducer, emptyFieldScheme);
    expect(renderer.create(wrapper.get(0)).toJSON()).toMatchSnapshot();
  });

  it("reders full alive field", () => {
    const wrapper = setupWrapper(reducer, emptyFieldScheme);
    expect(renderer.create(wrapper.get(0)).toJSON()).toMatchSnapshot();
  });

  it("renders empty field 4x5", () => {
    const wrapper = setupWrapper(reducer, emptyFieldScheme);

    const fieldSize = emptyFieldScheme[0].length * emptyFieldScheme.length;

    expect(
      wrapper.findWhere((el) => el.is(Cell) && el.props().isAlive === false)
        .length
    ).toBe(fieldSize);
  });

  it("renders full field 4x5", () => {
    const field = setupWrapper(reducer, fullFieldScheme);

    const fieldSize = fullFieldScheme[0].length * fullFieldScheme.length;

    expect(
      field.findWhere((el) => el.is(Cell) && el.props().isAlive === true).length
    ).toBe(fieldSize);
  });

  it("renders partial filled field, size 3*4", () => {
    const partialFilledField = [
      [false, true, true, true],
      [true, false, false, false],
      [true, true, false, true],
      [false, true, false, true],
    ];

    const [aliveCells, deadCells] = partialFilledField.reduce(
      (result, current) => {
        return [
          result[0] + current.filter((el) => el === true).length,
          result[1] + current.filter((el) => el === false).length,
        ];
      },
      [0, 0]
    );

    const field = setupWrapper(reducer, partialFilledField);

    expect(
      field.findWhere((el) => el.is(Cell) && el.props().isAlive === false)
        .length
    ).toBe(deadCells);

    expect(
      field.findWhere((el) => el.is(Cell) && el.props().isAlive === true).length
    ).toBe(aliveCells);

    expect(
      field.findWhere((el) => el.is(Cell) && el.props().isAlive === true).length
    ).not.toBe(1);
  });
});
