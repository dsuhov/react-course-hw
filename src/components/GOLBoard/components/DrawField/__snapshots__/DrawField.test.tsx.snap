// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field that draw cells reders full alive field 1`] = `
.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 45px;
  margin: 0 auto;
  border: 5px solid #cdc392;
  box-shadow: 0 0 10px #304c89;
  border-radius: 2px;
}

.emotion-0 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: 1px solid rgb(85,74,58);
  border-top: 1px solid rgb(85,74,58);
}

.emotion-1 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: none;
  border-top: 1px solid rgb(85,74,58);
}

.emotion-4 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: 1px solid rgb(85,74,58);
  border-top: none;
}

.emotion-5 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: none;
  border-top: none;
}

<div
  className="emotion-20"
  size={4}
>
  <div
    className="emotion-0"
    onClick={[Function]}
    x={0}
    y={0}
  />
  <div
    className="emotion-1"
    onClick={[Function]}
    x={1}
    y={0}
  />
  <div
    className="emotion-1"
    onClick={[Function]}
    x={2}
    y={0}
  />
  <div
    className="emotion-1"
    onClick={[Function]}
    x={3}
    y={0}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={1}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={1}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={1}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={1}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={2}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={2}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={2}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={2}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={3}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={3}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={3}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={3}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={4}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={4}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={4}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={4}
  />
</div>
`;

exports[`Field that draw cells reders full empty field 1`] = `
.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 45px;
  margin: 0 auto;
  border: 5px solid #cdc392;
  box-shadow: 0 0 10px #304c89;
  border-radius: 2px;
}

.emotion-0 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: 1px solid rgb(85,74,58);
  border-top: 1px solid rgb(85,74,58);
}

.emotion-1 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: none;
  border-top: 1px solid rgb(85,74,58);
}

.emotion-4 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: 1px solid rgb(85,74,58);
  border-top: none;
}

.emotion-5 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: none;
  border-top: none;
}

<div
  className="emotion-20"
  size={4}
>
  <div
    className="emotion-0"
    onClick={[Function]}
    x={0}
    y={0}
  />
  <div
    className="emotion-1"
    onClick={[Function]}
    x={1}
    y={0}
  />
  <div
    className="emotion-1"
    onClick={[Function]}
    x={2}
    y={0}
  />
  <div
    className="emotion-1"
    onClick={[Function]}
    x={3}
    y={0}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={1}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={1}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={1}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={1}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={2}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={2}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={2}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={2}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={3}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={3}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={3}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={3}
  />
  <div
    className="emotion-4"
    onClick={[Function]}
    x={0}
    y={4}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={1}
    y={4}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={2}
    y={4}
  />
  <div
    className="emotion-5"
    onClick={[Function]}
    x={3}
    y={4}
  />
</div>
`;
